### PR TITLE
Questions V1

### PR DESCRIPTION
**Pull request by bendjukastein**
Made 5 endpoints for the Question google cloud-run microservice that interface with our Firestore database:

Get all question Ids
{{baseUrl}}/api/questions

Get question title and content from question Id
{{baseUrl}}/api/questions/{{questionID}}

Get Questions Responses from question ID
{{baseUrl}}/api/questions/{{questionID}}

Get Q_Responses from Q_ID
GET {{baseUrl}}/api/questions/{{questionID}}/responses

Post question response
POST {{baseUrl}}/api/questions/{{questionID}}/responses

### CODE REVIEW

File: server.ts [hunk 1]
```diff
@@ -1,5 +1,15 @@
 import express, { Express } from "express";
 import cors from "cors";
+import { Firestore, QueryDocumentSnapshot, QuerySnapshot, Timestamp } from "@google-cloud/firestore";
+import { Question } from "./src/model/Question";
+import { QuestionResponse } from "./src/model/QuestionResponse";
+
+//TODO when hosting to cloud run:
+//Add your credentials to process.env, e.g. the ProjectId
+const firestore = new Firestore({
+  projectId: process.env.ProjectId,
```
Comment (kylesaburao) [inline server.ts:10]: prefer all caps PROJECT_ID for consistency so less likely to mess up during replication

Comment (kylesaburao) [reply server.ts:10]: since other env var is like that

File: server.ts [hunk 2]
```diff
@@ -8,45 +18,187 @@ app.use(cors({
...
+    title: string;
+    content: string;
   */
-
-  res.json("UNIMPLEMENTED");
-  // Return created UUID
+  // uuid left blank for now, generated in firestore.
+  const created: Timestamp = Timestamp.now()
+  const submissionData: Question = {
+    authorUUID: "123", // TODO: set proper authorUUID once users service is online
```
Comment (kylesaburao) [inline server.ts:58]: `authorID`

File: server.ts [hunk 3]
```diff
@@ -8,45 +18,187 @@ app.use(cors({
...
+    date: created, //date
+    title: req.body.title,//title
+    content: req.body.content //content
+  }
+  firestore.collection("questions").add(
+    submissionData
+  ).then((doc: any) => {
+    console.info('stored new doc', doc);
+    return res.status(200).send(doc);
```
Comment (kylesaburao) [inline server.ts:67]: don't send anything back, do what jarod did for blogs

File: server.ts [hunk 4]
```diff
@@ -8,45 +18,187 @@ app.use(cors({
...
+  });
 });
 
-// Get question thread
+// GET question title and content, given the question's id
+//params: id in param of request
 app.get("/api/questions/:questionUUID", (req, res) => {
-  res.json("UNIMPLEMENTED");
+  return firestore.collection("questions")
+    .doc(req.params.questionUUID)
```
Comment (kylesaburao) [inline server.ts:81]: `questionID`

ur own db would have to be adjusted

File: server.ts [hunk 5]
```diff
@@ -8,45 +18,187 @@ app.use(cors({
...
+      error: 'Unable to upload new question',
+      err
+    });
+  });
 });
 
-// Get question thread
+// GET question title and content, given the question's id
+//params: id in param of request
 app.get("/api/questions/:questionUUID", (req, res) => {
```
Comment (kylesaburao) [inline server.ts:79]: `questionID`

File: server.ts [hunk 6]
```diff
@@ -8,45 +18,187 @@ app.use(cors({
...
+        error: 'Unable to retrieve the Question',
+        err
+      });
+    });
 });
 
-// Respond to question
+//Get QuestionResponses
+//Given a Question id, returns the collection of related Responses
+app.get("/api/questions/:questionUUID/responses", (req, res) => {
```
Comment (kylesaburao) [inline server.ts:97]: `questionID`

File: server.ts [hunk 7]
```diff
@@ -8,45 +18,187 @@ app.use(cors({
...
+  const qResponseArray: QuestionResponse[] = [];
+  firestore.collection("QuestionResponses")
+    .where("questionUUID", "==", req.params.questionUUID)
```
Comment (kylesaburao) [inline server.ts:100]: `questionID`

File: server.ts [hunk 8]
```diff
@@ -8,45 +18,187 @@ app.use(cors({
...
+        return res.status(404).send({
+          error: 'Unable to find the questionResponses'
+        });
+      }
+      qResponses.forEach((qResponse: QueryDocumentSnapshot) => {
+        //push each questionResponse to the array we will return
+        const data = qResponse.data();
+        qResponseArray.push({
+          message: data.message,
+          questionUUID: data.questionUUID,
```
Comment (kylesaburao) [inline server.ts:113]: `questionID`

File: server.ts [hunk 9]
```diff
@@ -8,45 +18,187 @@ app.use(cors({
...
+          date: data.date,
+          authorUUID: data.authorUUID,
```
Comment (kylesaburao) [inline server.ts:115]: `authorID`

File: server.ts [hunk 10]
```diff
@@ -8,45 +18,187 @@ app.use(cors({
...
+        });
+      });
+      const responseData: string = JSON.stringify(qResponseArray);
+      console.log('send data in response:', responseData);
+      //return res.status(200).send(responseData); //uncomment if you only want the Array.
+      return res.status(200).send(
```
Comment (kylesaburao) [inline server.ts:121]: api spec only defines as return the list of responses

Comment (kylesaburao) [reply server.ts:121]: [{}, {}, ...]

File: server.ts [hunk 11]
```diff
@@ -8,45 +18,187 @@ app.use(cors({
...
-  */
-
-  res.json("UNIMPLEMENTED");
+  firestore.collection("questions")
+    .doc(req.params.questionUUID)
+    .get()
+    .then((doc: any) => {
+      if (doc.exists && doc.get("title")) {
+        //continue to post a Qresponse
+        const created: Timestamp = Timestamp.now();
```
Comment (kylesaburao) [inline server.ts:144]: use numbers and Date.now

File: server.ts [hunk 12]
```diff
@@ -8,45 +18,187 @@ app.use(cors({
...
+        const submissionData: QuestionResponse = {
+          questionUUID: req.params.questionUUID,
```
Comment (kylesaburao) [inline server.ts:146]: `questionID`

File: server.ts [hunk 13]
```diff
@@ -8,45 +18,187 @@ app.use(cors({
...
+          authorUUID: "123",// filler authorUUID for version 1
```
Comment (kylesaburao) [inline server.ts:147]: `authorID`

File: src/model/Question.ts [hunk 14]
```diff
@@ -1,7 +1,9 @@
+import { Timestamp } from "@google-cloud/firestore";
+
 export interface Question {
-    uuid: string;
+    uuid?: string;
```
Comment (kylesaburao) [inline src/model/Question.ts:4]: `id?: string`

File: src/model/Question.ts [hunk 15]
```diff
@@ -1,7 +1,9 @@
...
     authorUUID: string;
-    date: number;
+    date: Timestamp;
```
Comment (kylesaburao) [inline src/model/Question.ts:6]: `date: number`

File: src/model/QuestionResponse.ts [hunk 16]
```diff
@@ -1,7 +1,9 @@
+import { Timestamp } from "@google-cloud/firestore";
+
 export interface QuestionResponse {
-  uuid: string;
+  uuid?: string;
```
Comment (kylesaburao) [inline src/model/QuestionResponse.ts:4]: `id?: string`

File: src/model/QuestionResponse.ts [hunk 17]
```diff
@@ -1,7 +1,9 @@
...
   questionUUID: string;
   authorUUID: string;
-  date: number;
+  date: Timestamp;
```
Comment (kylesaburao) [inline src/model/QuestionResponse.ts:7]: `date: number`

**Review by kylesaburao — CHANGES_REQUESTED** [review]

File: server.ts (hunk 3 above)
Comment (kylesaburao) [reply server.ts:67]: `.send();`

File: server.ts (hunk 10 above)
Comment (kylesaburao) [reply server.ts:121]: directly as a list, not a list inside a object

File: server.ts (hunk 8 above)
Comment (kylesaburao) [reply server.ts:114]: `id:`

**Review by kylesaburao — CHANGES_REQUESTED** [review]

**Comment by kylesaburao** [comment]
every instance of `UUID` to `ID` plssss

File: server.ts (hunk 4 above)
Comment (bendjukastein) [reply server.ts:81]: I've edited my db to reflect the name change from uuid to id

**Review by kylesaburao — APPROVED** [review]
Merge once ur confirmed it works 100%

File: server.ts [hunk 18]
```diff
@@ -8,45 +18,183 @@ app.use(cors({
...
+ ******************************************************************************/
+// Firestore states field value length limit for indexed fields as 1,500 bytes
+// - UTF-8 chars take up 2 bytes each: therefore 750 character length limit
+// If you want to increase this limit, verify that we are using NON-INDEXED...
+// ...field values 
+// See: https://firebase.google.com/docs/firestore/quotas#limits
+const GCLOUD_STRING_LENGTH_LIMIT = 750;
+// DB_STR_LIMIT slightly shorter for safety
+const DB_STR_LIMIT = GCLOUD_STRING_LENGTH_LIMIT - 50;
+
```
Comment (jarod-f) [inline server.ts:32]: Unused constants here - would be good to remove these or make use of them when checking strings that are headed into the Firestore database

File: server.ts [hunk 19]
```diff
@@ -8,45 +18,183 @@ app.use(cors({
...
+    authorID: "123", // TODO: set proper authorID once users service is online
+    date: created, //date
+    title: req.body.title,//title
+    content: req.body.content //content
+  }
+  firestore.collection("questions").add(
+    submissionData
+  ).then((doc: any) => {
+    console.info('stored new doc', doc);
+    return res.status(200).send();
```
Comment (jarod-f) [inline server.ts:67]: This request code should be 201 to align with API

File: server.ts [hunk 20]
```diff
@@ -8,45 +18,183 @@ app.use(cors({
...
+  }).catch((err: any) => {
+    console.error(err);
+    return res.status(404).send({
```
Comment (jarod-f) [inline server.ts:70]: This request code should be 400 to align with API

File: server.ts [hunk 21]
```diff
@@ -8,45 +18,183 @@ app.use(cors({
...
+      error: 'Unable to upload new question',
+      err
+    });
```
Comment (jarod-f) [inline server.ts:73]: Should have empty `send()` body to align with API

File: server.ts [hunk 22]
```diff
@@ -8,45 +18,183 @@ app.use(cors({
...
-
-  res.json("UNIMPLEMENTED");
-  // Return created UUID
+  // id left blank for now, generated in firestore.
+  const created: number = Date.now();
+  const submissionData: Question = {
+    authorID: "123", // TODO: set proper authorID once users service is online
+    date: created, //date
+    title: req.body.title,//title
+    content: req.body.content //content
```
Comment (jarod-f) [inline server.ts:61]: Maybe we could trim strings here using the `DB_STR_LIMIT` from above? Just to make sure we're not going over the limit https://firebase.google.com/docs/firestore/quotas#limits

The limit would only apply in this case if it's an **indexed value** but I haven't verified what kind of values we're using - best to be on the safe side

File: server.ts [hunk 23]
```diff
@@ -8,45 +18,183 @@ app.use(cors({
...
+const DB_STR_LIMIT = GCLOUD_STRING_LENGTH_LIMIT - 50;
+
 app.use(express.json());
 app.use(express.urlencoded({ extended: true }));
 
+/******************************************************************************
+ *                                API Endpoints                               *
+ ******************************************************************************/
+
 app.use("/", (req, _, next) => {
```
Comment (jarod-f) [inline server.ts:40]: Might want to consider giving explicit type declartions to your endpoint objects throughout this file

```js
... (req: Request, _: Response, next: NextFunction) => ...
```

File: server.ts [hunk 24]
```diff
@@ -8,45 +18,183 @@ app.use(cors({
...
+        questionIDs.push(doc.id);
+      });
+      res.status(200).json(questionIDs);
+    }).catch((err: any) => {
+      console.error(err);
+      res.status(404).send({
+        error: 'Unable to retrieve the question ids',
+        err
+      });
+    });
```
Comment (jarod-f) [inline server.ts:197]: Might want to consider explicitly stating the types of variables throughout this endpoint
i.e. `questionIDs`, `data`, and `doc`?

Looks like you also imported the associated types already for some of these: `QueryDocumentSnapshot` & `QuerySnapshot`

File: server.ts [hunk 25]
```diff
@@ -8,45 +18,183 @@ app.use(cors({
...
-  res.json("UNIMPLEMENTED");
+  const questionIDs: any = [];
+  //query firestore for all question ids, returning them in an array
+  return firestore.collection("questions")
+    .get()
+    .then((data: any) => {
+      data.forEach((doc: any) => {
+        questionIDs.push(doc.id);
+      });
+      res.status(200).json(questionIDs);
```
Comment (jarod-f) [inline server.ts:190]: oh nice! I didn't know expresss had a `.json` send call for when you know your input is in JSON format!

File: src/model/QuestionResponse.ts [hunk 26]
```diff
@@ -1,7 +1,9 @@
+import { Timestamp } from "@google-cloud/firestore";
```
Comment (jarod-f) [inline src/model/QuestionResponse.ts:1]: Since we're going with number for date - I think we can remove this import

File: src/model/Question.ts [hunk 27]
```diff
@@ -1,6 +1,8 @@
+import { Timestamp } from "@google-cloud/firestore";
```
Comment (jarod-f) [inline src/model/Question.ts:1]: Since we're going with number for date - I think we can remove this import

File: server.ts [hunk 28]
```diff
@@ -1,5 +1,15 @@
 import express, { Express } from "express";
 import cors from "cors";
+import { Firestore, QueryDocumentSnapshot, QuerySnapshot, Timestamp } from "@google-cloud/firestore";
```
Comment (jarod-f) [inline server.ts:3]: Since we're going with number for date - I think we can remove this `Timestamp` import

File: server.ts [hunk 29]
```diff
@@ -8,45 +18,183 @@ app.use(cors({
...
+    .get()
+    .then((doc: any) => {
+      //HHH TODO: make sure Doc is of type Question
+      return res.status(200).send(doc);
+    }).catch((err: any) => {
+      console.error(err);
+      return res.status(404).send({
+        error: 'Unable to retrieve the Question',
+        err
+      });
```
Comment (jarod-f) [inline server.ts:91]: I think this should be a `400` with empty `send()` to match the API

File: server.ts [hunk 30]
```diff
@@ -8,45 +18,183 @@ app.use(cors({
...
-app.get("/api/questions/:questionUUID", (req, res) => {
-  res.json("UNIMPLEMENTED");
+// GET question title and content, given the question's id
+//params: id in param of request
+app.get("/api/questions/:questionID", (req, res) => {
+  return firestore.collection("questions")
+    .doc(req.params.questionID)
+    .get()
+    .then((doc: any) => {
+      //HHH TODO: make sure Doc is of type Question
```
Comment (jarod-f) [inline server.ts:84]: Could we maybe make an issue for this so we remember to get it done?

File: server.ts [hunk 31]
```diff
@@ -8,45 +18,183 @@ app.use(cors({
...
+app.get("/api/questions/:questionID/responses", (req, res) => {
+  const qResponseArray: QuestionResponse[] = [];
+  firestore.collection("QuestionResponses")
+    .where("questionID", "==", req.params.questionID)
+    .get()
+    .then((qResponses: QuerySnapshot) => {
+      if (qResponses.docs.length === 0) {
+        return res.status(404).send({
+          error: 'Unable to find the questionResponses'
+        });
```
Comment (jarod-f) [inline server.ts:106]: I think this should be a `400` with empty `send()` to match the API

File: server.ts [hunk 32]
```diff
@@ -8,45 +18,183 @@ app.use(cors({
...
+      });
+      const responseData: string = JSON.stringify(qResponseArray);
+      console.log('send data in response:', responseData);
+      return res.status(200).send(responseData);
+    }).catch((err: any) => {
+      console.error(err);
+      return res.status(404).send({
+        error: 'Unable to retrieve the Question',
+        err
+      });
```
Comment (jarod-f) [inline server.ts:127]: I think this should be a `400` with empty `send()` to match the API

File: server.ts [hunk 33]
```diff
@@ -8,45 +18,183 @@ app.use(cors({
...
+    .get()
+    .then((qResponses: QuerySnapshot) => {
+      if (qResponses.docs.length === 0) {
+        return res.status(404).send({
+          error: 'Unable to find the questionResponses'
+        });
+      }
+      qResponses.forEach((qResponse: QueryDocumentSnapshot) => {
+        //push each questionResponse to the array we will return
+        const data = qResponse.data();
```
Comment (jarod-f) [inline server.ts:110]: Might want to add an explicit type here?

File: server.ts [hunk 34]
```diff
@@ -8,45 +18,183 @@ app.use(cors({
...
+          authorID: "123",// filler authorID for version 1
+          date: created,
+          message: req.body.message,
+        };
+        firestore
+          .collection("QuestionResponses")
+          .add(submissionData)
+          .then((doc: any) => {
+            console.info("stored new QuestionResponse", doc);
+            return res.status(200).send();
```
Comment (jarod-f) [inline server.ts:152]: I think this should be response code `201` to align with API spec

File: server.ts [hunk 35]
```diff
@@ -8,45 +18,183 @@ app.use(cors({
...
+          })
+          .catch((err: any) => {
+            console.error(err);
+            return res.status(404).send({
+              error: "unable to upload new QuestionResponse",
+              err,
+            });
```
Comment (jarod-f) [inline server.ts:159]: I think this should be response code `400` with empty body to align with API spec

File: server.ts [hunk 36]
```diff
@@ -8,45 +18,183 @@ app.use(cors({
...
+          });
+      } else {
+        return res.status(404).send({
+          error: "Invalid Question to add response to",
+        });
```
Comment (jarod-f) [inline server.ts:164]: I think this should be response code `400` with empty body to align with API spec

File: server.ts [hunk 37]
```diff
@@ -8,45 +18,183 @@ app.use(cors({
...
+      }
+    })
+    .catch((err: any) => {
+      console.error(err);
+      return res.status(404).send({
+        error:
+          "Unable to validate Question. Try checking that your question ID exists",
+        err,
+      });
```
Comment (jarod-f) [inline server.ts:173]: I think this should be response code `400` with empty body to align with API spec

File: server.ts [hunk 38]
```diff
@@ -8,45 +18,183 @@ app.use(cors({
...
+        //continue to post a Qresponse
+        const created: number = Date.now();
+        const submissionData: QuestionResponse = {
+          questionID: id,
+          authorID: "123",// filler authorID for version 1
+          date: created,
+          message: req.body.message,
+        };
+        firestore
+          .collection("QuestionResponses")
```
Comment (jarod-f) [inline server.ts:148]: I think for consistency with your other calls, these two lines should be one line

File: server.ts [hunk 39]
```diff
@@ -8,45 +18,183 @@ app.use(cors({
...
+        error: 'Unable to retrieve the Question',
+        err
+      });
+    });
+});
 
-  res.json("UNIMPLEMENTED");
+// POST QuestionResponse
+app.post("/api/questions/:questionID/responses", (req, res) => {
+  const id = req.params.questionID;
```
Comment (jarod-f) [inline server.ts:133]: `questionID` might be a more descriptive variable name here

**Review by jarod-f — CHANGES_REQUESTED** [review]
The collections `questions` and `QuestionResponse` have inconsistent casing - I think it might be good to change them to be consistent with each other
- Additionally, they could maybe be stored in named constants at the top of the file instead of hard-coded into each call i.e. `const QUESTIONS_COLLECTION = "Questions"`

File: server.ts (hunk 22 above)
Comment (bendjukastein) [reply server.ts:61]: agreed, I'll use pull in your blog helper function if that's alright by you

**Pull request merged**